### PR TITLE
feat: add support for native checksums

### PR DESCRIPTION
**Pull request by lauzadis**
*Issue #, if available:*

*Description of changes:*
Adds support for SHA-1, SHA-256, CRC32, and CRC32C via CRT bindings.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-crt-kotlin/api/aws-crt-kotlin.api
Comment (ianbotsf) [inline aws-crt-kotlin/api/aws-crt-kotlin.api:1]: **Question**: Why was this file deleted?

File: aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/ShaNative.kt [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+/**
+ * A typealias to CRT's aws_<checksum_algorithm>_new functions (i.e aws_sha1_new, aws_sha256_new))
+ */
+internal typealias InitializeHashFn = (
+    allocator: CValuesRef<aws_allocator>?,
+) -> CPointer<aws_hash>?
+
+public open class ShaNative(
+    public val initializeFn: InitializeHashFn,
+) {
```
Comment (ianbotsf) [inline aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/ShaNative.kt:20]: **Question**: Why `public`?

File: aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/ShaNative.kt [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    // aws_hash_finalize
+    public fun digest(): ByteArray {
+        val output: CPointer<aws_byte_buf> = aws_mem_calloc(
+            Allocator.Default,
+            1.convert(),
+            sizeOf<aws_byte_buf>().convert(),
+        )?.reinterpret() ?: throw CrtRuntimeException("aws_mem_calloc() aws_byte_buf")
+        aws_byte_buf_init(output, Allocator.Default.allocator, 10_000U)
```
Comment (ianbotsf) [inline aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/ShaNative.kt:44]: **Question**: Why 10K?

**Review by ianbotsf — APPROVED** [review]

File: aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/ShaNative.kt (hunk 2 above)
Comment (lauzadis) [reply aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/ShaNative.kt:44]: Ah this was overlooked, I will reduce it to the minimum size necessary

File: aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/ShaNative.kt (hunk 1 above)
Comment (lauzadis) [reply aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/ShaNative.kt:20]: If it's not `public`, there are compiler errors because Sha1Native and Sha256Native are exposing the internal class

File: aws-crt-kotlin/api/aws-crt-kotlin.api
Comment (lauzadis) [reply aws-crt-kotlin/api/aws-crt-kotlin.api:1]: I deleted all of the `.api` files and re-ran, `./gradlew apiDump`, it didn't re-create this one. I run `./gradlew apiCheck` and it passes, so I'm not sure why this was created.

I will add a backlog task to start running `./gradlew apiCheck` as a GitHub Actions CI in this repository since it seems to be missing.

File: aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/ShaNative.kt [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        }
+    }
+
+    // aws_hash_finalize
+    public fun digest(): ByteArray {
+        val output: CPointer<aws_byte_buf> = aws_mem_calloc(
+            Allocator.Default,
+            1.convert(),
+            sizeOf<aws_byte_buf>().convert(),
+        )?.reinterpret() ?: throw CrtRuntimeException("aws_mem_calloc() aws_byte_buf")
```
Comment (0marperez) [inline aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/ShaNative.kt:43]: **Question**: What is this exception trying to say?

File: aws-crt-kotlin/common/test/aws/sdk/kotlin/crt/util/DigestTest.kt [hunk 4]
```diff
@@ -5,12 +5,11 @@
...
 package aws.sdk.kotlin.crt.util
 
-import kotlin.test.Ignore
+import aws.sdk.kotlin.crt.CrtTest
 import kotlin.test.Test
 import kotlin.test.assertEquals
 
-class DigestTest {
-    @Ignore // FIXME Enable when Kotlin/Native implementation is complete
+class DigestTest : CrtTest() {
```
Comment (0marperez) [inline aws-crt-kotlin/common/test/aws/sdk/kotlin/crt/util/DigestTest.kt:12]: **Question**: Could we test Sha1 here as well? Is that tested somewhere else?

File: aws-crt-kotlin/native/test/aws/sdk/kotlin/crt/util/hashing/ShaNativeTest.kt [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    fun testMultipleUpdatesWithOffset() {
+        // algorithm -> hash("uh") + hash("huh")
+        val tests = listOf(
+            (Sha1Native() to "c093098e7a1a8ec547e82a399d63f331760e9a57"),
+            (Sha256Native() to "b474fb57f69c1b5aa00aa27ad0b36c03fe9915cc497344d6a55a86b2e9bd1b72"),
+        )
+
+        tests.forEach { (hash, expected) ->
+            val data = "ahuhi".encodeToByteArray()
+            hash.update(data, 2, 2) // uh
```
Comment (0marperez) [inline aws-crt-kotlin/native/test/aws/sdk/kotlin/crt/util/hashing/ShaNativeTest.kt:40]: **Question**: Maybe I'm misunderstanding but shouldn't this be `hash.update(data, 2, 3)`?

File: aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/ShaNative.kt (hunk 3 above)
Comment (lauzadis) [reply aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/ShaNative.kt:43]: It's referencing the name of the CRT function which failed (`aws_mem_calloc()`) specifically when trying to allocate memory for an `aws_byte_buf`.

We have a similar style across the rest of the Kotlin/Native implementation, it helps identify which C function failed.

File: aws-crt-kotlin/native/test/aws/sdk/kotlin/crt/util/hashing/ShaNativeTest.kt (hunk 5 above)
Comment (lauzadis) [reply aws-crt-kotlin/native/test/aws/sdk/kotlin/crt/util/hashing/ShaNativeTest.kt:40]: The argument order is `data`, `offset`, `length`, so to get the string "uh" from "ahuhi", it's an offset of 2 with a length of 2. Offset is zero-indexed so a value of 2 means start at the third character.

File: aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/ShaNative.kt (hunk 1 above)
Comment (aajtodd) [reply aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/ShaNative.kt:20]: Then we should find a way to hide those details no? It could be as simple as introducing an interface.

File: aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/ShaNative.kt [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+    private fun initializeHash() = checkNotNull(initializeFn(Allocator.Default.allocator)) {
+        "failed to initialize hash"
+    }
+}
+
+/**
+ * SHA-1 hash function implemented using bindings to CRT
+ */
+public class Sha1Native : ShaNative(::aws_sha1_new)
```
Comment (aajtodd) [inline aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/ShaNative.kt:68]: **fix**: Naming of _types_ shouldn't include platform generally (at least not public types).

**Review by aajtodd — CHANGES_REQUESTED** [review]
> CRC32 support comes from KMP's platform.zlib.crc32 and will be implemented directly in smithy-kotlin

Why? Can we not get CRC32 directly from CRT? If we use `platform.zlib` we end up with a direct zlib dependency required on the host OS. If we can get it from CRT we should.

**Comment by lauzadis** [comment]
> > CRC32 support comes from KMP's platform.zlib.crc32 and will be implemented directly in smithy-kotlin
> 
> Why? Can we not get CRC32 directly from CRT? If we use `platform.zlib` we end up with a direct zlib dependency required on the host OS. If we can get it from CRT we should.

@aajtodd We can definitely get CRC32 from CRT. Kotlin's documentation made me think [`platform.zlib`](https://kotlinlang.org/docs/native-platform-libs.html#popular-native-libraries) is available on all targets but that might not be true?

**Comment by aajtodd** [comment]
> > > CRC32 support comes from KMP's platform.zlib.crc32 and will be implemented directly in smithy-kotlin
> > 
> > 
> > Why? Can we not get CRC32 directly from CRT? If we use `platform.zlib` we end up with a direct zlib dependency required on the host OS. If we can get it from CRT we should.
> 
> @aajtodd We can definitely get CRC32 from CRT. [Kotlin's documentation made me think](https://kotlinlang.org/docs/native-platform-libs.html#popular-native-libraries) `platform.zlib` is available on all targets but that might not be true?

They are but

>  Kotlin/Native compiler automatically detects which of the platform libraries have been accessed and automatically links the needed libraries.
> On the other hand, the platform libs in the distribution are merely just wrappers and bindings to the native libraries. That means the native libraries themselves (.so, .a, .dylib, .dll etc) should be installed on the machine.

I think I'd err on the side of just using CRT if it's easy enough to add.

File: aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/ShaNative.kt (hunk 1 above)
Comment (lauzadis) [reply aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/ShaNative.kt]: Addressed by adding a `HashFunction` interface

File: aws-crt-kotlin/build.gradle.kts [hunk 7]
```diff
@@ -177,6 +177,7 @@ kotlin {
     val xcrun = "/usr/bin/xcrun"
 
     tasks.register<Exec>("bootIosSimulatorDevice") {
+        isIgnoreExitValue = true
```
[the remark this answers is not in the document]
Comment (0marperez) [reply aws-crt-kotlin/build.gradle.kts:180]: It might be worth mentioning it in the TLS negotiation errors PR and linking to this one or creating a separate PR

File: aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/ShaNative.kt [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+            aws_byte_cursor_from_array(it.addressOf(offset), length.convert())
+        }
+        awsAssertOpSuccess(aws_hash_update(hash, inputCursor)) {
+            "aws_hash_update"
+        }
+    }
+
+    // aws_hash_finalize
+    override fun digest(): ByteArray {
+        val output: CPointer<aws_byte_buf> = aws_mem_calloc(
```
Comment (aajtodd) [inline aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/ShaNative.kt:57]: **suggestion**. You can probably do this without an allocation using `ByteArray` directly by constructing the byte buf yourself.

Something like:
```kotlin
val output = ByteArray(32)
output.usePinned { pinned ->
     val buf = cValueOf<aws_byte_buf> {
          capacity = ...
         len = 0
         buffer = addressOf(0)
      }
}
```

**Review by aajtodd — APPROVED** [review]
Minor suggestion. Looks good.
